### ISSUE TITLE
re-evaluate amount of reserved slop space

### ISSUE THREAD

**Issue by fling-** [openzfs/zfs#3014]
zpool reports 2% free but free space available for the dataset is 3210M which is ~0,1%

```
studio ~ # uname -a  
Linux studio 3.17.4-hardened-r1-gnu_fling #2 SMP PREEMPT Tue Dec 9 13:28:46 OMST 2014 x86_64 AMD Phenom(tm) II X6 1100T Processor AuthenticAMD GNU/Linux

studio ~ # cat /sys/module/{spl,zfs}/version
0.6.3-1.1
0.6.3-1.1

studio ~ # zpool list -v plot10
NAME   SIZE  ALLOC   FREE    CAP  DEDUP  HEALTH  ALTROOT
plot10  2,72T  2,67T  46,6G    98%  1.00x  ONLINE  -
  sdi1  2,72T  2,67T  46,6G         -

studio ~ # zfs list -r -t filesystem,snapshot,volume plot10
NAME                  USED  AVAIL  REFER  MOUNTPOINT
plot10               2,67T  3,13G   136K  none
plot10/burstcoin     2,67T  3,13G   136K  none
plot10/burstcoin/10  2,67T  3,13G  2,67T  /var/lib/lxc/burst/var/lib/plot/10

studio ~ # zpool get all plot10
NAME    PROPERTY               VALUE                  SOURCE
plot10  size                   2,72T                  -
plot10  capacity               98%                    -
plot10  altroot                -                      default
plot10  health                 ONLINE                 -
plot10  guid                   12610495349026519243   default
plot10  version                -                      default
plot10  bootfs                 -                      default
plot10  delegation             on                     default
plot10  autoreplace            off                    default
plot10  cachefile              -                      default
plot10  failmode               wait                   default
plot10  listsnapshots          off                    default
plot10  autoexpand             off                    default
plot10  dedupditto             0                      default
plot10  dedupratio             1.00x                  -
plot10  free                   46,6G                  -
plot10  allocated              2,67T                  -
plot10  readonly               off                    -
plot10  ashift                 12                     local
plot10  comment                -                      default
plot10  expandsize             0                      -
plot10  freeing                0                      default
plot10  leaked                 0                      default
plot10  feature@async_destroy  enabled                local
plot10  feature@empty_bpobj    active                 local
plot10  feature@lz4_compress   enabled                local

studio ~ # zfs get all plot10
NAME    PROPERTY              VALUE                      SOURCE
plot10  type                  filesystem                 -
plot10  creation              Вс ноя 23  5:00 2014  -
plot10  used                  2,67T                      -
plot10  available             3,13G                      -
plot10  referenced            136K                       -
plot10  compressratio         1.00x                      -
plot10  mounted               no                         -
plot10  quota                 none                       default
plot10  reservation           none                       default
plot10  recordsize            128K                       default
plot10  mountpoint            none                       local
plot10  sharenfs              off                        default
plot10  checksum              off                        local
plot10  compression           off                        local
plot10  atime                 off                        local
plot10  devices               on                         default
plot10  exec                  on                         default
plot10  setuid                on                         default
plot10  readonly              off                        default
plot10  zoned                 off                        default
plot10  snapdir               hidden                     default
plot10  aclinherit            restricted                 default
plot10  canmount              on                         default
plot10  xattr                 sa                         local
plot10  copies                1                          default
plot10  version               5                          -
plot10  utf8only              on                         -
plot10  normalization         formD                      -
plot10  casesensitivity       sensitive                  -
plot10  vscan                 off                        default
plot10  nbmand                off                        default
plot10  sharesmb              off                        default
plot10  refquota              none                       default
plot10  refreservation        none                       default
plot10  primarycache          metadata                   local
plot10  secondarycache        metadata                   local
plot10  usedbysnapshots       0                          -
plot10  usedbydataset         136K                       -
plot10  usedbychildren        2,67T                      -
plot10  usedbyrefreservation  0                          -
plot10  logbias               latency                    default
plot10  dedup                 off                        default
plot10  mlslabel              none                       default
plot10  sync                  standard                   default
plot10  refcompressratio      1.00x                      -
plot10  written               136K                       -
plot10  logicalused           2,67T                      -
plot10  logicalreferenced     15K                        -
plot10  snapdev               hidden                     default
plot10  acltype               off                        default
plot10  context               none                       default
plot10  fscontext             none                       default
plot10  defcontext            none                       default
plot10  rootcontext           none                       default
plot10  relatime              off                        default

studio ~ # zpool history plot10
History for 'plot10':
2014-11-23.05:01:14 zfs create plot10/burstcoin
2014-11-23.05:01:28 zfs create plot10/burstcoin/10 -o mountpoint=/var/lib/lxc/burst/var/lib/plot/10
2014-12-09.12:36:21 zfs set primarycache=metadata plot10
2014-12-09.12:36:32 zfs set secondarycache=metadata plot10

```

**Comment by AndCycle** [comment]
http://cuddletech.com/blog/pivot/entry.php?id=1013
that's an intentional design, 

if you fill up COW filesystem, it will go into a corner that it can't get out,
because COW need free space to do any operation, if you fill it, you are done

**Comment by behlendorf** [comment]
Right, internally ZFS reserves a small amount of space to ensure it always has working room.  We could perhaps lie about this more convincingly in the reported capacity to avoid this kind of confusion.

**Comment by BjoKaSH** [comment]
just run into a similar problem while upgrading a box from 0.6.4.1 to 0.6.5.4.

A small pool (648 GB total, physical disk size) that had 55.2 GB (9%) unallocated space according to `zpool list`  showed about 5.3 GB free in `zfs list` under 0.6.4.1. After the upgrade it shows 0 bytes free in `zfs list` and is unusable, i.e. effectively read-only.  I can not even create a final snapshot to transfer the pool's content to some larger pool.

So: What exactly is "**a small amount of space** to ensure it always has working room"?
10% seems a lot and 55 GB either.  I'd have expected something like max( 500 \* max( blocksize_of_each_dataset ) \* num_vdev, min( 5% of pool, 4GB*num_vdev ) )

Also, what is this "working room" used for when you even can't do a `zfs create snapshot` to, for example, do a zfs send afterwards?

Lastly, is there a tunable (module parameter, sysctl) to (temporary) lower the reservation, such that the pool can be used (e.g. to replicate elsewhere)?  A `sysctl -a` and a look through `/sys/module/zfs/parameter` didn't showed an obvious candidate.

@behlendorf Not sure if "lying more convincingly" is a good idea, but what should be avoided is, that an upgrade of ZoL renders a working pool into a half-dead one.

Reconsidering the "lying", I'd think the only convincing way would be to set aside a few GB right at pool creation, and of course correctly account for them, i.e. a `zpool list` just after a create should show the reserved space in the "ALLOC" column. (And probably updating the reservation similar to above made-up formula when datasets with larger blocksize are created.)

**Comment by kernelOfTruth** [comment]
@BjoKaSH 
>Default reserved space was increased from 1.6% to 3.2% of total pool capacity. This default percentage can be controlled through the new spa_slop_shift module option, setting it to 6 will restore the previous percentage.
https://github.com/zfsonlinux/zfs/releases

**Comment by GregorKopka** [comment]
On a pool with 1 raidz vdev and 5 2TB drives
```
filer ~ $ zfs list tank
NAME    USED  AVAIL  REFER  MOUNTPOINT
tank   6,98T  29,9G   345K  /tank

filer ~ $ zpool list tank
NAME    SIZE  ALLOC   FREE  EXPANDSZ   FRAG    CAP  DEDUP  HEALTH  ALTROOT
tank   9,06T  8,74T   327G     16,0E      -    96%  1.00x  ONLINE  -
```
this hides ~150GB (worst case used: every write from now on will be 1 block only so raidz parity overhead will be 100%) of useable space.

This seems to be excessive.

@behlendorf I don't see a reason why using a percentage of the total pool size should be neccecary, especially since the minimum hidden size is hardcoded as 32MB.

How about we limit the size returned by spa_get_slop_space (https://github.com/zfsonlinux/zfs/blob/887d1e60ef1f20a1b13e7c7b0d208f10b13b9cbe/module/zfs/spa_misc.c#L1605) to something like a maximum of 1GB (or whatever reasonable)?

**Comment by behlendorf** [comment]
@GregorKopka you can decrease this value locally on your system if you absolutely need to use that space.  However, it's hard to put an exact upper bound on the minimum required free space needed to avoid certain pathological situations.  Reserving this space additionally helps ensure a functional level of performance even with a 99+% full pool.

**Comment by GregorKopka** [comment]
@behlendorf Would it be feasible to turn the setting into a pool-level property/tuneable?

Asking since the spa_slop_shift being system wide this seems to be a bit unpractical as all systems I run have at least a small root pool and decondary big data pools, it would be helpful to be able to tune this per pool and not per system.

Still confused about the absolute minimum being 4 orders of magnitude lower than what's eaten away from the 8TB example I posted above...

**Comment by GregorKopka** [comment]
https://github.com/zfsonlinux/zfs/issues/4117#issuecomment-165490366 hints spa_slop_shift dosn't work as expected:

>It talks about 1.6% and 3.6% of total pool capacity. I have a pool with 15TB of capacity. Alternating spa_slop_shift option between 5 and 6 produces a change of almost 3TB of available space, and that's almost 20% though :. So I'm missing something for sure.

**Comment by GregorKopka** [comment]
@ahrens as you seem to be the primary author on this, could you please share the reasoning behind spa_slop_shift scaling linear with the pool size? Could you please shed some light on

What other operations could call _dsl_sync_task(ZFS_SPACE_CHECK_NONE)_ apart from the 'e.g. "zfs destroy"' that is listed in the spa_slop_shift comment? 
Can't an absolute worst case of _free space needed_ for these operations be determined or does their demand really scale with pool size? 
Under what conditions could ZFS ever need >10GB (just to pull a number from thin air) of reserved free space so it won't paint itself into a corner permanently? 

Simply put: What argument could be made against a fixed spa_max_slop (possibly accompanied by, if needed, an overriding spa_min_slop_per_vdev - should additional vdevs increase the need for reserved space in a pool) to put a limit on the absolute amount of space being made unavailable?

**Comment by ahrens** [comment]
@GregorKopka For purposes of keeping enough free space to accommodate writes to the MOS, we might be able to safely put an upper bound on the slop space.  I don't think that this question has been studied recently, either empirically or theoretically.

It's tricky to figure out what the limit should be - you're inherently saying, "well we'll *never* do this many things in one TXG".  As time goes on, the maximum scale tends to increase (doing more things at once, increasing the amount of slop space requried), and pool sizes also tend to increase (increasing the amount of slop space).  While this connection is admittedly loose, it serves to future-proof this safeguard.  If we had hard coded a maximum slop space when we began developing ZFS in 2001 (hard drives were ~100GB!), it probably would not be large enough today.

Note that the slop space is not just for "zfs destroy".  The slop space must be large enough for all writes to the MOS, i.e. all sync tasks and spacemaps.  The ZFS_SPACE_CHECK_* determines which operations are allowed to continue when we're already 97% full.

The slop space also serves to prevent performance from becoming awful.  Typically by the time the pool is 97% allocated, performance is so bad that it's preferable to stop and get a human to intervene.  For this purpose linearly scaling makes sense.  But I could see arguments that we should let folks degrade their performance if they choose to.

**Comment by nathanshearer** [comment]
I have just started to consider using zfs on some of my systems, and on a test system, the slop space allocation is becoming somewhere ridiculous. A 90TB raidz2 is reserving over 2TB of slop space! That is a bit absurd to require 2TB for reserved space to improve on-disk allocation, data alignment, etc.

Consider a system with even more disks. Growing a reserved memory region like this is causing massive overhead and wasted resources. On peta- and exa-scale operations this fixed-percentage allocation gets even more wasteful.

Can I suggest allowing this parameter to be tuneable? I can see situations where a large and small slop space is appropriate, but for most datasets a hard-coded fixed-percentage is probably not the right size.

**Comment by behlendorf** [comment]
For reference there is a proposal in PR #11023 to apply a reasonable maximum value.

**Comment by behlendorf** [comment]
Resolved by #11023.
